src/main/java/jenkins/branch/MultiBranchProjectViewHolder.java

```diff
@@ -0,0 +1,285 @@
+    /**
+     * The primary view name.
+     */
+    @GuardedBy("this")
+    private transient String primaryView = null;
+
+    /**
+     * Constructor.
+     *
+     * @param owner the owning {@link MultiBranchProject}.
+     */
+    public MultiBranchProjectViewHolder(MultiBranchProject<?, ?> owner) {
+        super();
+        this.owner = owner;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @NonNull
+    @Override
+    public synchronized List<View> getViews() {
+        if (owner.getItems().isEmpty()) {
+            // when there are no branches nor pull requests to show, switch to the special welcome view
+            return Collections.singletonList(owner.getWelcomeView());
+        }
+        if (views == null) {
+            List<View> views = new ArrayList<>();
+            for (SCMHeadCategory c : SCMHeadCategory.collectAndSimplify(owner.getSCMSources()).values()) {
+                views.add(new ViewImpl(owner, c));
+                if (c.isUncategorized()) {
+                    primaryView = c.getName();
+                }
+            }
+            this.views = views;
+        }
+        return views;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setViews(@NonNull List<? extends View> views) {
+        // ignore
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized String getPrimaryView() {
+        if (owner.getItems().isEmpty()) {
+            // when there are no branches nor pull requests to show, switch to the special welcome view
+            return BaseEmptyView.VIEW_NAME;
+        }
+        if (primaryView == null) {
+            getViews(); // will set primaryView for us
+        }
+        return primaryView;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setPrimaryView(@CheckForNull String name) {
+        // ignore
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isPrimaryModifiable() {
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @NonNull
+    @Override
+    public ViewsTabBar getTabBar() {
+        if (tabBar == null) {
+            tabBar = new LocalizedViewsTabBar();
+        }
+        return tabBar;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setTabBar(@NonNull ViewsTabBar tabBar) {
+        // ignore
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isTabBarModifiable() {
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void invalidateCaches() {
+        views = null;
+        primaryView = null;
+    }
+
+    /**
+     * A custom category specific view.
+     */
+    @Restricted(NoExternalUse.class)
+    public static class ViewImpl extends ListView {
+
+        /**
+         * The category that this view selects.
+         */
+        @NonNull
+        private final SCMHeadCategory category;
```

Comment:
field is hidden, impossible to get category for the given view

Reply:
Pr for getter welcome. Quit sniping, file PRs to improve if you think they are required, i’ll Discuss potential changes, but sniping on random commits is not winning you friends here